Frameworks/CoreText/DWriteWrapper_CTFont.mm

```diff
@@ -208,6 +208,173 @@ HRESULT _DWriteCreateFontFaceWithFontDescriptor(CTFontDescriptorRef fontDescript
+            if (CFStringCompare(name, fontName, kCFCompareCaseInsensitive) == kCFCompareEqualTo) {
+                matchingFonts.emplace_back(font);
+            }
+        }
+    }
+
+    if (!matchingFonts.empty() || forceMatch) {
+        fonts = std::move(matchingFonts);
+    }
+
+    return S_OK;
+}
+
+HRESULT __FilterFontsByKeyAndValue(std::vector<ComPtr<IDWriteFont>>& fonts, CFStringRef key, CFTypeRef value, bool forceMatch) {
+    if (CFStringCompare(key, kCTFontFamilyNameAttribute, 0) == kCFCompareEqualTo) {
+        return __FilterFontsByFamilyName(fonts, static_cast<CFStringRef>(value), forceMatch);
+    } else if (CFStringCompare(key, kCTFontNameAttribute, 0) == kCFCompareEqualTo) {
+        return __FilterFontsByInformationalString(fonts,
+                                                  static_cast<CFStringRef>(value),
+                                                  DWRITE_INFORMATIONAL_STRING_POSTSCRIPT_NAME,
+                                                  forceMatch);
+    } else if (CFStringCompare(key, kCTFontDisplayNameAttribute, 0) == kCFCompareEqualTo) {
+        return __FilterFontsByInformationalString(fonts,
+                                                  static_cast<CFStringRef>(value),
+                                                  DWRITE_INFORMATIONAL_STRING_FULL_NAME,
+                                                  forceMatch);
+    } else if (CFStringCompare(key, kCTFontStyleNameAttribute, 0) == kCFCompareEqualTo) {
+        return __FilterFontsByStyleName(fonts, static_cast<CFStringRef>(value), forceMatch);
+    } else if (CFStringCompare(key, kCTFontTraitsAttribute, 0) == kCFCompareEqualTo) {
+        __FilterFontsByTraits(fonts, static_cast<CFDictionaryRef>(value), forceMatch);
+        return S_OK;
+    } else {
+        // Other attributes are currently unsupported by CTFont
+        UNIMPLEMENTED();
+        return S_FALSE;
+    }
+}
+
+#pragma endregion // Helper Filter Methods
+
+/**
+ * Creates a CFArray of CTFontDescriptors values for all fonts that match the given attributes best and completely match mandatoryKeys
+ */
+HRESULT _DWriteCreateMatchingFontDescriptors(CFDictionaryRef attributes, CFSetRef mandatoryKeys, CFArrayRef* matchingNames) {
+    *matchingNames = nullptr;
+
+    // Get list of all possible fonts
+    std::vector<ComPtr<IDWriteFont>> fonts;
+    RETURN_IF_FAILED(_DWriteGetAllFonts(fonts));
+
+    if (mandatoryKeys) {
+        // Filter by the mandatory keys first
+        CFIndex count = CFSetGetCount(mandatoryKeys);
+        std::vector<CFStringRef> keys(count);
+        CFSetGetValues(mandatoryKeys, (const void**)keys.data());
+        for (auto& key : keys) {
+            if (!CFDictionaryContainsKey(attributes, key)) {
+                // Mandatory key does not exist, so no fonts can match
+                return S_OK;
+            }
+
+            auto value = woc::MakeAutoCF<CFTypeRef>(CFDictionaryGetValue(attributes, key));
```

Comment:
just use an unguarded local. no need to retain it and then release it across a single sequence point.